### PR TITLE
Feat: introduce `Safebox` contracts

### PR DESCRIPTION
**Pull request by amusingaxl**
A safebox for digital assets.

- The `owner` can request funds to be sent to the `recipient`.
- The `custodian` can deny the request within a predefined time period (24h).
- After the timelock, anyone can execute the withdrawal.
- The `custodian` cooperation is required whenever the `owner` wants to update the `recipient`.

### CODE REVIEW

**Review by dbkbali — APPROVED** [review]
Looks good to me - based on our discussions

**Review by nduchak — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  - [Change the `recipient` address](#change-the-recipient-address)
+
+## Overview
+
+There are 3 "roles" in this contract:
+
+1. `owner`: has full control of how much and when it can send assets to `recipient`.
+2. `custodian`: approves changes in the `recipient` address made by the `owner`.
+3. `recipient`: receives assets from the safebox.
+
```
Comment (dbkbali) [inline README.md:23]: Need to update the readme for the new variables and the addition of the VatLike interface

**Review by dbkbali — CHANGES_REQUESTED** [review]

File: src/Safebox.sol [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        emit Withdraw(token, amount);
+    }
+
+    /**
+     * @dev Handles `transfer` for tokens with non-standard ERC-20 implementations.
+     * See https://github.com/d-xo/weird-erc20
+     * @param to The destination address.
+     * @param amount The amount to be transfered.
+     */
+    function _safeTransfer(address token, address to, uint256 amount) internal {
```
Comment (gbalabasquer) [inline src/Safebox.sol]: I'd reduce the verbosity of the code just using a simple `transfer`. I don't see Maker interacting with a token that doesn't `revert` and this contract is behind governance setting it up for a specific purpose. This is only useful for contracts that are open to any token.
At least I'd remove the internal function and just put the code in `withdraw` (same for `deposit`) as there is not reuse of it. But honestly I'd just keep it simple with the regular `transfer`.

File: src/Safebox.sol [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"recipient"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "recipient") {
```
Comment (rockyfour) [inline src/Safebox.sol:255]: This should be blocked when VAT is not live.

File: src/Safebox.sol [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+     * @param usr The user address
+     * @return Whether `usr` is a ward or not.
+     */
+    function isCustodian(address usr) external view returns (bool) {
+        return can[usr] == 1;
+    }
+
+    /**
+     * @notice Approves the change in the recipient.
+     * @dev Reverts if `previousRecipient` has not been set or if `_recipient` does not match it.
```
Comment (rockyfour) [inline src/Safebox.sol]: `previousRecipient` => `pendingRecipient` ?

File: src/Safebox.sol [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity ^0.8.16;
+
+/**
+ * @author amusingaxl
+ * @title A safebox for digital assets.
+ * @notice
+ * - The `owner` is in full control of how much and when it can send assets to `recipient`.
+ *   - If MakerDAO governance ever executes an Emergency Shutdown, anyone can send assets to `recipient`.       This prevents assets being stuck in this contract when the governance smart contract is no longer operational.
```
Comment (rockyfour) [inline src/Safebox.sol]: This line is really long

File: src/Safebox.sol [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+     * @notice `usr` was granted custodian access.
+     * @param usr The user address.
+     */
+    event Hope(address indexed usr);
+    /**
+     * @notice The owner withdrawn assets from the safebox.
+     * @param token The token withdrawn.
+     * @param amount The amount withdrawn.
+     */
+    event Withdraw(address indexed token, uint256 amount);
```
Comment (rockyfour) [inline src/Safebox.sol]: This should probably emit the recipient as well.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  - [Deposit assets](#deposit-assets)
+  - [Withdraw assets](#withdraw-assets)
+  - [Change the `recipient` address](#change-the-recipient-address)
+
+## Overview
+
+There are 3 "roles" in this contract:
+
+1. `owner`: has full control of how much and when it can send assets to `recipient`.
+2. `custodian`: approves changes in the `recipient` address made by the `owner`.
```
Comment (rockyfour) [inline README.md]: mmm namewise -  this entity is not really storing/custodying the tokens, it can only block changing the recipient.

Comment (amusingaxl) [reply README.md]: It has to be named like this for legal reasons.

File: src/Safebox.sol [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+ * - The `custodian` cooperation is required whenever the `owner` wants to update the `recipient`.
+ */
+contract Safebox {
+    /// @notice MCD Vat module.
+    VatLike public immutable vat;
+
+    /// @notice Addresses with owner access on this contract. `wards[usr]`
+    mapping(address => uint256) public wards;
+    /// @notice Addresses with custodian access on this contract. `can[usr]`
+    mapping(address => uint256) public can;
```
Comment (rockyfour) [inline src/Safebox.sol]: If custodian is effectively immutable and can only be set in the constructor why is this mapping needed?

Comment (gbalabasquer) [reply src/Safebox.sol]: Yeah this is a good observation. Let's just make an immutable variable and remove the mapping.

File: src/Safebox.sol (hunk 3 above)
Comment (gbalabasquer) [reply src/Safebox.sol:255]: ++

File: src/Safebox.sol (hunk 8 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Coinbase requested the ability to be able to rotate their keys, so this will be needed. I'm adding the functionality right now.

**Comment by amusingaxl** [comment]
Coinbase is not comfortable with the idea that MakerDAO governance could add or remove custodians at will. Since this will not be accessible by governance, I decided to rename the `can` mapping to `custodians`. Any custodian can add or remove other custodians to this contract.

File: src/Safebox.sol (hunk 3 above)
Comment (amusingaxl) [reply src/Safebox.sol:255]: Done

File: src/Safebox.sol [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    //////////////////////////////////*/
+
+    /**
+     * @notice Withdraws ERC20-like tokens from this contract.
+     * @dev Anyone can call this function after MakerDAO governance executes an Emergency Shutdown.
+     * @param token The token to be withdrawn.
+     * @param amount The amount of tokens.
+     */
+    function withdraw(address token, uint256 amount) external {
+        require(wards[msg.sender] == 1 || vat.live() == 0, "Safebox/not-ward");
```
Comment (amusingaxl) [inline src/Safebox.sol:153]: I'm tempted to refactor this into:

```suggestion
        if (vat.live() == 1) {
            require(wards[msg.sender] == 1, "Safebox/not-ward");
        }
```

This is less gas efficient, but clearer on the intent.

File: src/Safebox.sol (hunk 5 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Visual glitch on my editor. Already fixed.

File: src/Safebox.sol (hunk 4 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Fixed.

File: src/Safebox.sol (hunk 6 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Done.

Also added the `sender` to the `Deposit` event.

File: src/Safebox.sol (hunk 3 above)
Comment (amusingaxl) [reply src/Safebox.sol:174]: Fixed.

File: src/Safebox.sol (hunk 2 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Makes sense, removed.

Comment (amusingaxl) [reply src/Safebox.sol]: Another thing I considered was to make `token` a constructor parameter instead of a parameter for `withdraw`.

As it is, the Safebox can be seen as multi-token, but there will probably never be a case for it to hold more than 1 asset.

File: README.md (hunk 1 above)
Comment (amusingaxl) [reply README.md:23]: Addressed in the section below.

File: src/Safebox.sol [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+     * @return Whether `usr` is a ward or not.
+     */
+    function isOwner(address usr) external view override returns (bool) {
+        return wards[usr] == 1;
+    }
+
+    /**
+     * @notice Check if an address has `custodian` access on this contract.
+     * @param usr The user address.
+     * @return Whether `usr` is a ward or not.
```
Comment (rockyfour) [inline src/Safebox.sol]: `usr` => `custodian`

File: src/Safebox.sol [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    function addCustodian(address usr) external override onlyCustodian {
+        custodians[usr] = 1;
+        emit AddCustodian(usr);
+    }
+
+    /**
+     * @notice Removes a custodian from this contract.
+     * @param usr The user address.
+     */
+    function removeCustodian(address usr) external override onlyCustodian {
```
Comment (rockyfour) [inline src/Safebox.sol]: Note that as with `wards` behavior, also here if there is more than one custodian this will allow any custodian to remove any other.
So need to make sure this is wanted.

File: src/Safebox.sol [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        emit RemoveCustodian(usr);
+    }
+
+    /**
+     * @notice Approves the change in the recipient.
+     * @dev Reverts if `pendingRecipient` has not been set or if `_recipient` does not match it.
+     * @param _recipient The new recipient being approved.
+     */
+    function approveChangeRecipient(address _recipient) external override {
+        require(custodians[msg.sender] == 1, "Safebox/not-custodian");
```
Comment (rockyfour) [inline src/Safebox.sol]: Should this be replaced with the `onlyCustodian` modifier?

File: src/Safebox.sol [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+
+        wards[_owner] = 1;
+        emit Rely(_owner);
+
+        custodians[_custodian] = 1;
+        emit AddCustodian(_custodian);
+
+        vat = VatAbstract(_vat);
+        token = GemAbstract(_token);
+        recipient = _recipient;
```
Comment (rockyfour) [inline src/Safebox.sol:131]: Consider emitting a `RecipientChange` event (and maybe changing the event name to `RecipientSet` or something similar).

File: src/Safebox.sol [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+     * @param _vat The MCD vat module.
+     * @param _token The ERC-20 token to be held in this contract.
+     * @param _owner The safebox owner.
+     * @param _custodian The safebox custodian.
+     * @param _recipient The recipient for tokens in the safebox.
+     */
+    constructor(address _vat, address _token, address _owner, address _custodian, address _recipient) {
+        require(_recipient != address(0), "Safebox/invalid-recipient");
+
+        wards[_owner] = 1;
```
Comment (rockyfour) [inline src/Safebox.sol:125]: Note that common behavior for MakerDao contracts is to only rely msg.sender in the constructor.

File: src/Safebox.sol [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+        token.transfer(recipient, amount);
+        emit Withdraw(recipient, amount);
+    }
+
+    /**
+     * @notice Deposits tokens into this contract.
+     * @param amount The amount of tokens.
+     */
+    function deposit(uint256 amount) external override {
```
Comment (rockyfour) [inline src/Safebox.sol]: Do you plan to use this function for depositing from the swap/output conduit?
Seems like current code passes the funds by specifying the recipient of `buyGem`.

File: src/Safebox.sol (hunk 11 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Yes, that's the expected feature.

File: src/Safebox.sol (hunk 13 above)
Comment (amusingaxl) [reply src/Safebox.sol:131]: All other events have "verb-like" structure (i.e.: `Rely`, `Deny`, `AddCustodian`), I'm gonna use `SetRecipient` for this.

File: src/Safebox.sol (hunk 14 above)
Comment (amusingaxl) [reply src/Safebox.sol:125]: Yeah, we are aware of that, however, since Coinbase will be handling the deployment, we want to avoid them having to `deny` themselves afterwards.

File: src/Safebox.sol (hunk 15 above)
Comment (amusingaxl) [reply src/Safebox.sol]: You are right, I'm gonna remove that.
The idea was to keep track of `Deposit` events, but we can do the same by tracking `Push` from the `RwaSwapOutputConduit` contract.

File: src/Safebox.sol (hunk 14 above)
Comment (amusingaxl) [reply src/Safebox.sol:125]: Just to clarify, we are gonna provide them the parameters for deployment and then verify the contract after it's deployed. If there is a mismatch, the verification will fail and we will know.

Also the variables are public, so they could be easily checked after deployment & verification.

File: src/Safebox.sol (hunk 9 above)
Comment (The-Arbiter) [reply src/Safebox.sol:153]: If `vat.live()` is 0 it'll skip this check. If you want to refactor this just make sure that all cases are still checked.

File: src/Safebox.sol [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+ * - The `custodian` cooperation is required whenever the `owner` wants to update the `recipient`.
+ */
+contract Safebox is SafeboxLike {
+    /// @notice MCD Vat module.
+    VatAbstract public immutable override vat;
+    /// @notice The ERC-20 token to be held in this contract.
+    GemAbstract public immutable override token;
+
+    /// @notice Addresses with owner access on this contract. `wards[usr]`
+    mapping(address => uint256) public override wards;
```
Comment (The-Arbiter) [inline src/Safebox.sol]: Spacing
```suggestion
    mapping (address => uint256) public override wards;
```

File: src/Safebox.sol [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+    /// @notice Addresses with custodian access on this contract. `custodians[usr]`
+    mapping(address => uint256) public override custodians;
```
Comment (The-Arbiter) [inline src/Safebox.sol]: Spacing
```suggestion
    mapping (address => uint256) public override custodians;
```

File: src/Safebox.sol [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+    /// @notice The recipient for the tokens held in this contract.
+    address public override recipient;
+    /// @notice Reference to the new recipient when it is to be changed.
+    address public override pendingRecipient;
+
+    modifier auth() {
```
Comment (The-Arbiter) [inline src/Safebox.sol]: Not crazy on the layout of code but it's subjective...

Maybe we should do it like autoline? (https://github.com/makerdao/dss-auto-line/blob/master/src/DssAutoLine.sol) where the rely deny and auth are together in the code? Up to you

File: src/Safebox.sol [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    /**
+     * @param _vat The MCD vat module.
+     * @param _token The ERC-20 token to be held in this contract.
+     * @param _owner The safebox owner.
+     * @param _custodian The safebox custodian.
+     * @param _recipient The recipient for tokens in the safebox.
+     */
+    constructor(address _vat, address _token, address _owner, address _custodian, address _recipient) {
```
Comment (The-Arbiter) [inline src/Safebox.sol:117]: Any reason for the ordering of the params here? Doesn't seem to follow any specific logic - up to you here but I would order it either alphabetically or based on order of use within the constructor (owner custodian recipient vat token). Again up to you

File: src/Safebox.sol [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+        require(_recipient != address(0), "Safebox/invalid-recipient");
+
+        wards[_owner] = 1;
+        emit Rely(_owner);
+
+        custodians[_custodian] = 1;
+        emit AddCustodian(_custodian);
```
Comment (The-Arbiter) [inline src/Safebox.sol:129]: Why not `RelyCustodian` to keep it in line with existing naming conventions?

File: src/Safebox.sol [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+    /*//////////////////////////////////
+            MakerDAO Interfaces
+    //////////////////////////////////*/
+
+    /**
+     * @notice Grants `usr` owner access to this contract.
+     * @param usr The user address.
+     */
+    function rely(address usr) external override auth {
```
Comment (The-Arbiter) [inline src/Safebox.sol]:
```suggestion
    function rely(address usr) external auth {
```

File: src/Safebox.sol [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` owner access from this contract.
+     * @param usr The user address.
+     */
+    function deny(address usr) external override auth {
```
Comment (The-Arbiter) [inline src/Safebox.sol]:
```suggestion
    function deny(address usr) external auth {
```

File: src/Safebox.sol [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"recipient"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external override auth {
```
Comment (The-Arbiter) [inline src/Safebox.sol]: Make this `public` and no override?

```suggestion
    function file(bytes32 what, address data) public auth {
```

File: README.md [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+- Goerli: `TBD`
+
+## Usage
+
+### Deposit assets
+
+There are 2 possible ways of making a deposit into the `Safebox`:
+
+1. Use the `Safebox` address as the `to` address of an ERC-20 compatible `transfer` transaction.
+2. Call the `deposit`<sup>[1]</sup> method from the `Safebox` contract:
```
Comment (The-Arbiter) [inline README.md]: If this is removed we should remove this :)

**Review by The-Arbiter — COMMENTED** [review]
Hey guys, here is my brief analysis of SafeBox. I have extensive experience with escrows so I am reviewing it from the perspective of an escrow in "Maker style" :)
Since I am reviewing this in isolation, context may be missing with regards to some comments, however I have made various points as documented below:

### TLDR
- 'Dumb depositing' ERC20 poses risks. I will go into this in the **Attack Vectors** section at the end.
- Everything is `override` and I really don't think it should be (also nothing is `virtual` in `SafeBoxLike.sol`)
- Some of the names could be changed to be more like regular Daiwanese, especially the `addCustodian` stuff
- The structure of the 2 step ownership makes checking for `address(0)` edge cases mandatory however allows for smart contract wallets / recipients. This is a contextual thing, I want to confirm that this is intended.
- `IsOwner` and `IsCustodian` seem redundant unless I'm missing some context here.
- Comments and NATSPEC could do with trimming down depending on how `dss`-ish you want the code to look; this resembles how I wrote my own code prior to adapting to the `dss` style so this is just a general comment.

### Documentation
- This looks good, just needs updates when it's all complete.
- I would appreciate some notes on why `address(0)` is checked for (i.e. to allow `pendingRecipient` to work only if it's initialised)

### SafeBox.sol

- AGPL license OK
- All interfaces used
- Not immediately clear to me how the Abstracts are used with the `external view` return functions in `SafeBoxLike.sol` but this is probably my lack of knowledge (have never tried to just return an imported interface before). Knowledge gap here so someone else needs to fill this in and make sure it's OK.
- `VatAbstract` and `GemAbstract` OK - do we really need to use dss interfaces for only 3 interfaces? Personally I would use `VatLike` and `GemLike`.
- Mappings OK - fixed spacing
- Recipient and pending recipient vars OK 
- Auth mapping OK
- `onlyCustodian` code OK - modifier name seems a bit out of place since it sits next to `auth` - not a big issue
- Constructor OK
  - Rely owner OK
  - Add custodian OK
    I don't like `AddCustodian` as a name when we have `rely` already used as a keyword here. 
  - Set recipient OK
  - Set vat OK
  - Set gem OK
    I would change the param order to match either when they're used (sequentially) or make it alphabetically ordered, but that's me.
- `withdraw` has an attack vector explained below but the code is OK.
- `rely` - why is this override?
- `deny` - same issue
- `file` is OK - I made it `public` and not `override` but this is optional as it varies between implementations
  The `address(0)` check is needed to stop issues with `pendingRecipient` and this should be noted in documentation.
- `isOwner` - why do we need this? Can't they just query `wards[usr]`? I don't see a need for this (I may be missing context?) but the code is OK.
- `isCustodian` - same as `isOwner`
- `addCustodian` OK but should be called `relyCustodian`
- `removeCustodian` OK but should be called `denyCustodian`
- `approveChangeRecipient` - so I have some notes on this:
  Existing implementation seems OK as it is. The benefit of it is that it doesn't rely on a call from the recipient's address. The drawback is that we have to ensure that the recipient isn't `address(0)` for this to be possible.
  [OZ](https://github.com/heldersepu/openzeppelin-contracts/blob/master/contracts/access/Ownable2Step.sol) has a 2 step transfer method however that requires that the recipient calls the accept function which works for EOAs but not things like smart contract wallets. In the end this depends on the usecase and I may be missing the context here.
- No `deposit` function is present anymore (the previous deposit function looked OK though)

### Attack Vectors

Assume that I am a malicious member of governance and decide to steal the funds of the SafeBox. I can use CREATE2crunch or another similar tool to determine the salt for a contract address which resembles an ERC20 token's address, then deploy my own ERC20 there which looks like the real thing, however, my `transfer` function is malicious.
While a `deposit` function would only transfer in tokens with address `token` (meaning that `deposit` and `withdraw` keeps a 'closed loop' such that the deposited token is the one being withdrawn), depositing directly to the contract would allow this kind of attack to occur if there was someone malicious responsible for the SafeBox deployment.

**Comment by rockyfour** [comment]
@amusingaxl can you please point us to the code of the input conduit with the support for post shutdown collateral distribution (basically the intended `recipient` of `SafeBox`)?

File: src/Safebox.sol (hunk 16 above)
Comment (amusingaxl) [reply src/Safebox.sol]: This is handled by `prettier-plugin-solidity`.

File: src/Safebox.sol (hunk 18 above)
Comment (amusingaxl) [reply src/Safebox.sol]: We're using CES layout for this code.

File: src/Safebox.sol (hunk 19 above)
Comment (amusingaxl) [reply src/Safebox.sol:117]: We're trying to logically group them:

MCD core modules > General contracts > Roles for this contract.

File: src/Safebox.sol (hunk 20 above)
Comment (amusingaxl) [reply src/Safebox.sol:129]: We're refraining from using Daiwanese for functions that are exclusively accessible to Coinbase.

File: src/Safebox.sol (hunk 9 above)
Comment (amusingaxl) [reply src/Safebox.sol:153]: That is the expected behavior.

This is an escape hatch in case of Emergency Shutdown: we will allow anyone to withdraw the funds from this contract into the `recipient`.

File: src/Safebox.sol (hunk 21 above)
Comment (amusingaxl) [reply src/Safebox.sol]: `override` is here to ensure it's implementing the same method declared in the  `SafeboxLike` interface.

File: src/Safebox.sol (hunk 23 above)
Comment (amusingaxl) [reply src/Safebox.sol]: Why `public` instead of `external`?

**Comment by amusingaxl** [comment]
Thank you for the thorough review @The-Arbiter.

Just providing some context:

We are between a rock and a hard place in a sense that we want to use the regular Daiwanese where it's possible, however there was some push back on that from the Coinbase side.

This contract is going to be operated by both MakerDAO Governance and Coinbase in a collaborative manner, so our strategy is trying to segregate the interfaces that each party will interact with. That's why we kept things like `wards`, `file`, etc. for the MakerDAO-facing methods, while the part that will mostly be accessed by Coinbase have "more common" names.

Contrary to most MCD contract out there, a ward is not able to add or remove "custodian" privileges, only another custodian can do so. That was explicitly requested by Coinbase, because they were not comfortable with the idea of MakerDAO Governance being freely able to replace them.

Now answering to some of your comments:

> Everything is override and I really don't think it should be (also nothing is virtual in SafeBoxLike.sol)

In previous Solidity version you were required to `override` functions declared in interfaces, looks like I forgot this behavior changed in Solidity [0.8.8](https://github.com/ethereum/solidity/blob/develop/Changelog.md#088-2021-09-27).

> Some of the names could be changed to be more like regular Daiwanese, especially the addCustodian stuff

As explained above, we are using Daiwanese only for MakerDAO-facing features, as this was a requirement from Coinbase.

> The structure of the 2 step ownership makes checking for address(0) edge cases mandatory however allows for smart contract wallets / recipients. This is a contextual thing, I want to confirm that this is intended.

For this specific deal, the recipient is going to be a smart contract (a `RwaSwapInputConduit2` instance to be more precise). I don't think adding a block for EOAs brings much in terms of security, because changing a recipient requires collaboration between both parties. The chances of having this being set to an unintended address is pretty slim, since changes have to go through the due legal process.

> IsOwner and IsCustodian seem redundant unless I'm missing some context here.

These are mainly for Coinbase's use. While the `custodians` mapping has a "common" name, they felt the need of having something closer to the legal language "owner", so we added a `isOwner` as a wrapper for `wards`. For consistency, we decided to add `isCustodian`, but the latter can be removed if PE feels strongly against it.

> Comments and NATSPEC could do with trimming down depending on how dss-ish you want the code to look; this resembles how I wrote my own code prior to adapting to the dss style so this is just a general comment.

Our understanding is that this is not a Core MCD contract, so it would be fine to use CES coding convention here. Again, it's not something that would be a blocker for us, however considering the Coinbase requirements, I find it very hard to make it 100% compliant anyways.

> I would appreciate some notes on why address(0) is checked for (i.e. to allow pendingRecipient to work only if it's initialised)

Noted! Will update the docs.

> Not immediately clear to me how the Abstracts are used with the external view return functions in SafeBoxLike.sol but this is probably my lack of knowledge (have never tried to just return an imported interface before). Knowledge gap here so someone else needs to fill this in and make sure it's OK.

From an ABI perspective, those types are simply translated to `address`. However, we usually store contract references with their actual type to make them easier to consume. It's out of the scope of this review, but the contract to enable the recovery of funds after Emergency Shutdown will use the `SafeboxLike` interface.

> VatAbstract and GemAbstract OK - do we really need to use dss interfaces for only 3 interfaces? Personally I would use VatLike and GemLike.

Those interfaces are pretty much set in stone at this point. We import the specific files directly, so it wouldn't be a problem of bloating contract source verification for instance. We could make some copy-pasta from `dss-interfaces` and remove the methods not used though if you think that's the best way.

> `onlyCustodian` code OK - modifier name seems a bit out of place since it sits next to auth - not a big issue

I've been back and forth with the layout of this contract a couple of times TBH 😅.
I'll probably move everything accessible only to coinbase to their designated section.

> Existing implementation seems OK as it is. The benefit of it is that it doesn't rely on a call from the recipient's address. The drawback is that we have to ensure that the recipient isn't address(0) for this to be possible.

Yes, this is by design. Recipient is not meant to play an active role here because.

---

Regarding the possible attack vector:

> Assume that I am a malicious member of governance and decide to steal the funds of the SafeBox. I can use CREATE2crunch or another similar tool to determine the salt for a contract address which resembles an ERC20 token's address, then deploy my own ERC20 there which looks like the real thing, however, my transfer function is malicious.
> While a deposit function would only transfer in tokens with address token (meaning that deposit and withdraw keeps a 'closed loop' such that the deposited token is the one being withdrawn), depositing directly to the contract would allow this kind of attack to occur if there was someone malicious responsible for the SafeBox deployment.

Wouldn't this be solved by having the diligence of checking the `token` value matches exactly the expected one after deployment **before** the onboarding spell?

We can even safeguard against this in the spell itself by having a sanity check like:

```solidity
require(address(safebox.token()) == chainlog.getAddress("<TOKEN>"), "Invalid safebox token");
```

Any other asset sent to this contract other than `token` itself is locked there forever.

**Comment by amusingaxl** [comment]
> @amusingaxl can you please point us to the code of the input conduit with the support for post shutdown collateral distribution (basically the intended `recipient` of `SafeBox`)?

I'm finishing the required changes and will post the link to it here as soon as I'm done.

**Comment by amusingaxl** [comment]
@The-Arbiter @rockyfour here's the input conduit that is going to be used as `recipient` in this case: https://github.com/makerdao/mip21-toolkit/pull/27/files#diff-42d74a78f57f1a5daa3af319c8bdbec3e7850b836cd393fb7caa0b3ada65a0e2

File: src/Safebox.sol [hunk 25]
```diff
@@ -0,0 +1,303 @@
...
+
+        emit RequestWithdrawal(msg.sender, amount);
+    }
+
+    /**
+     * @notice Cancels a withdrawal request.
+     */
+    function cancelWithdrawal() external auth {
+        require(requestedWithdrawalTime > 0 && requestedWithdrawalAmount > 0, "Safebox/no-pending-withdrawal");
+
```
Comment (0xdecr1pto) [inline src/Safebox.sol:160]: Should we disallow to cancel withdrawal is timelock already passed?
```
 require(requestedWithdrawalTime + WITHDRAWAL_TIMELOCK > block.timestamp, "Safebox/timelock-passed");
```
